app/plugin_db_driver.go

```diff
@@ -0,0 +1,259 @@
+	stMut   sync.Mutex
+	stMap   map[string]driver.Stmt
+	rowsMut sync.Mutex
+	rowsMap map[string]driver.Rows
+}
+
+func NewDriverImpl(s *Server) *DriverImpl {
+	return &DriverImpl{
+		s:       s,
+		connMap: make(map[string]*sql.Conn),
+		txMap:   make(map[string]driver.Tx),
+		stMap:   make(map[string]driver.Stmt),
+		rowsMap: make(map[string]driver.Rows),
+	}
+}
+
+func (d *DriverImpl) Conn() (string, error) {
+	conn, err := d.s.sqlStore.GetMaster().Db.Conn(context.Background())
+	if err != nil {
+		return "", err
+	}
+	connID := model.NewId()
+	d.connMut.Lock()
+	d.connMap[connID] = conn
+	d.connMut.Unlock()
+	return connID, nil
+}
+
+func (d *DriverImpl) ConnPing(connID string) error {
+	d.connMut.RLock()
+	conn := d.connMap[connID]
+	d.connMut.RUnlock()
+	return conn.Raw(func(innerConn interface{}) error {
+		return innerConn.(driver.Pinger).Ping(context.Background())
+	})
+}
+
+func (d *DriverImpl) ConnQuery(connID, q string, args []driver.NamedValue) (_ string, err error) {
+	var rows driver.Rows
+	d.connMut.RLock()
+	conn := d.connMap[connID]
+	d.connMut.RUnlock()
+	err = conn.Raw(func(innerConn interface{}) error {
+		rows, err = innerConn.(driver.QueryerContext).QueryContext(context.Background(), q, args)
+		return err
+	})
+	if err != nil {
+		return "", err
+	}
+
+	rowsID := model.NewId()
+	d.rowsMut.Lock()
+	d.rowsMap[rowsID] = rows
+	d.rowsMut.Unlock()
+
+	return rowsID, nil
+}
+
+func (d *DriverImpl) ConnExec(connID, q string, args []driver.NamedValue) (_ plugin.ResultContainer, err error) {
+	var res driver.Result
+	var ret plugin.ResultContainer
+	d.connMut.RLock()
+	conn := d.connMap[connID]
+	d.connMut.RUnlock()
+	err = conn.Raw(func(innerConn interface{}) error {
+		res, err = innerConn.(driver.ExecerContext).ExecContext(context.Background(), q, args)
+		return err
+	})
+	if err != nil {
+		return ret, err
+	}
+
+	ret.LastID, ret.LastIDError = res.LastInsertId()
+	ret.RowsAffected, ret.RowsAffectedError = res.RowsAffected()
+
+	return ret, nil
+}
+
+func (d *DriverImpl) ConnClose(connID string) error {
+	d.connMut.Lock()
+	err := d.connMap[connID].Close()
+	delete(d.connMap, connID)
+	d.connMut.Unlock()
+
+	return err
+}
+
+func (d *DriverImpl) Tx(connID string, opts driver.TxOptions) (_ string, err error) {
+	var tx driver.Tx
+	d.connMut.RLock()
+	conn := d.connMap[connID]
+	d.connMut.RUnlock()
+	err = conn.Raw(func(innerConn interface{}) error {
+		tx, err = innerConn.(driver.ConnBeginTx).BeginTx(context.Background(), opts)
+		return err
+	})
+	if err != nil {
+		return "", err
+	}
+
+	txID := model.NewId()
+	d.txMut.Lock()
+	d.txMap[txID] = tx
+	d.txMut.Unlock()
+	return txID, nil
+}
+
+func (d *DriverImpl) TxCommit(txID string) error {
+	d.txMut.Lock()
+	defer d.txMut.Unlock()
+	err := d.txMap[txID].Commit()
+	delete(d.txMap, txID)
+	return err
+}
+
+func (d *DriverImpl) TxRollback(txID string) error {
+	d.txMut.Lock()
+	defer d.txMut.Unlock()
+	err := d.txMap[txID].Rollback()
+	delete(d.txMap, txID)
+	return err
+}
```

Comment:
Wondering if we could do the Commit/Rollback calls outside of lock.

Reply:
Yep, definitely. I need to change to RWMutexes too.